libraries/chain/webassembly/crypto.cpp

```diff
@@ -250,4 +251,105 @@ namespace eosio { namespace chain { namespace webassembly {
       return return_code::success;
    }
 
+   void interface::bls_g1_add(span<const char> op1, span<const char> op2, span<char> result) const
+   {
+       bls12_381::g1 a = bls12_381::g1::fromJacobianBytesLE({reinterpret_cast<const uint8_t*>(op1.data()), 144}, false, true);
+       bls12_381::g1 b = bls12_381::g1::fromJacobianBytesLE({reinterpret_cast<const uint8_t*>(op2.data()), 144}, false, true);
+       bls12_381::g1 c = a.add(b);
+       c.toJacobianBytesLE({reinterpret_cast<uint8_t*>(result.data()), 144}, true);
+   }
+
+   void interface::bls_g2_add(span<const char> op1, span<const char> op2, span<char> result) const
+   {
+       bls12_381::g2 a = bls12_381::g2::fromJacobianBytesLE({reinterpret_cast<const uint8_t*>(op1.data()), 288}, false, true);
+       bls12_381::g2 b = bls12_381::g2::fromJacobianBytesLE({reinterpret_cast<const uint8_t*>(op2.data()), 288}, false, true);
+       bls12_381::g2 c = a.add(b);
+       c.toJacobianBytesLE({reinterpret_cast<uint8_t*>(result.data()), 288}, true);
+   }
+
+   void interface::bls_g1_mul(span<const char> point, span<const char> scalar, span<char> result) const
+   {
+       bls12_381::g1 a = bls12_381::g1::fromJacobianBytesLE({reinterpret_cast<const uint8_t*>(point.data()), 144}, false, true);
+       std::array<uint64_t, 4> b = bls12_381::scalar::fromBytesLE<4>({reinterpret_cast<const uint8_t*>(scalar.data()), 32});
+       bls12_381::g1 c = a.mulScalar(b);
+       c.toJacobianBytesLE({reinterpret_cast<uint8_t*>(result.data()), 144}, true);
+   }
+
+   void interface::bls_g2_mul(span<const char> point, span<const char> scalar, span<char> result) const
+   {
+       bls12_381::g2 a = bls12_381::g2::fromJacobianBytesLE({reinterpret_cast<const uint8_t*>(point.data()), 288}, false, true);
+       std::array<uint64_t, 4> b = bls12_381::scalar::fromBytesLE<4>({reinterpret_cast<const uint8_t*>(scalar.data()), 32});
+       bls12_381::g2 c = a.mulScalar(b);
+       c.toJacobianBytesLE({reinterpret_cast<uint8_t*>(result.data()), 288}, true);
+   }
+
+   void interface::bls_g1_exp(span<const char> points, span<const char> scalars, const uint32_t n, span<char> result) const
+   {
+       std::vector<bls12_381::g1> pv;
+       std::vector<std::array<uint64_t, 4>> sv;
+       pv.reserve(n);
+       sv.reserve(n);
+       for(uint32_t i = 0; i < n; i++)
+       {
+           bls12_381::g1 p = bls12_381::g1::fromJacobianBytesLE({reinterpret_cast<const uint8_t*>(points.data() + i*144), 144}, false, true);
```

Comment:
Before using the `span<>`s passed as arguments, need to validate that the size of the span is what is required.

Consider, for example, a contract currently with 64KB of memory calling `bls_g1_exp(64KB-8, 4, ..., n=1, ...)`. The host function validation wrappers will allow this call because the `points` span is 4 bytes long and within the 0 to 64KB-1 valid range of memory. However, `bls12_381::g1::fromJacobianBytesLE({reinterpret_cast<const uint8_t*>(points.data() + i*144), 144}` then goes on to use 144 bytes starting from 64KB-8. This will access invalid WASM memory leading to undefined behavior (what will likely happen _in this case_ and _at this point of time_ is a recovered SEGV which leaks the memory allocated by the `std::vector<>::reserve()` above).

Host functions must ensure they only access memory within the passed `span<>` ranges. So instead, need to validate that `points.size() == n*144`. Consider, for example, this host function
https://github.com/AntelopeIO/leap/blob/ba919c794165a9702df7c2ed6b09ac85ee64bd74/libraries/chain/webassembly/crypto.cpp#L119-L126
notice how if `op1` is not 64 bytes an error is returned. Since `bls_g1_exp()` does not return anything, the only reasonable option for `bls_g1_exp` may be to assert that `points.size() == n*144` (i.e. fail the transaction).

This same guidance applies to `result` as well. Notice that the above `alt_bn128_add` example checks that `result` is _at least_ 64 bytes. That may not be ideal and may have been an oversight in the original implementation -- I would suggest enforcing that the `result` here is exactly 144 bytes.

(this feedback applies to all the added host functions)

Reply:
Thanks for the detailed explanation and example. Much appreciated! I was expecting the execution to fail in those cases but didn't consider potential memory leaks. Of course, these checks are very important.

Updates in my most recent commit:

- added checks for all `span<>` ranges
- make use of `return_code::success`/`return_code::failure` as in BN host functions
- also see related changes in `cdt` (proper error propagation back to smart contract and then handling with `eosio::check`, again, exactly as in BN host functions)
- fixed indentation of tabs (3 ws instead of 4)